File: packages/brmem/src/brmem/content_limits.py
"""Ingestion-time guardrails for branch-memory entries.

`brmem put` is intended for text artifacts (plan documents, objective state,
short notes). Two cheap checks at the door keep stray binaries and oversized
files out of the ref pack, where they would otherwise become hard-to-reclaim
storage weight once a brmem ref pinned them.

Helpers here are pure (`bytes -> str | None`); the caller wraps the returned
fragment with `source_file` context and the `-f / --force` override hint.
"""

from __future__ import annotations

from typing import Final

MAX_ENTRY_BYTES: Final = 1 << 20  # 1 MiB
BINARY_SNIFF_BYTES: Final = 8 * 1024  # 8 KiB


def check_entry_size(raw: bytes) -> str | None:
    if len(raw) > MAX_ENTRY_BYTES:
        return f"is {format_bytes(len(raw))}; brmem entries are capped at 1 MiB"
    return None


def check_entry_not_binary(raw: bytes) -> str | None:
    head = raw[:BINARY_SNIFF_BYTES]
    offset = head.find(b"\x00")
    if offset == -1:
        return None
    return f"appears to be binary (NUL byte at offset {offset})"


def format_bytes(n: int) -> str:
    if n < 1024:
        return f"{n} B"
    if n < 1024 * 1024:
        return f"{n // 1024} KiB"
    if n < 1024**3:
        return f"{n / (1024 * 1024):.1f} MiB"
    return f"{n / (1024**3):.1f} GiB"

File: ts/packages/public/infra/brmem/src/content-limits.ts
export const MAX_ENTRY_BYTES = 1 << 20;
export const BINARY_SNIFF_BYTES = 8 * 1024;

export function checkEntrySize(bytes: Uint8Array): string | undefined {
	if (bytes.byteLength <= MAX_ENTRY_BYTES) return undefined;
	return `is ${formatBytes(bytes.byteLength)}; Branch Memory Entries are capped at 1 MiB`;
}

export function checkEntryNotBinary(bytes: Uint8Array): string | undefined {
	const limit = Math.min(bytes.byteLength, BINARY_SNIFF_BYTES);
	for (let index = 0; index < limit; index += 1) {
		if (bytes[index] === 0) return `appears to be binary (NUL byte at offset ${index})`;
	}
	return undefined;
}

export function formatBytes(n: number): string {
	if (n < 1024) return `${n} B`;
	if (n < 1024 * 1024) return `${Math.floor(n / 1024)} KiB`;
	if (n < 1024 ** 3) return `${(n / (1024 * 1024)).toFixed(1)} MiB`;
	return `${(n / 1024 ** 3).toFixed(1)} GiB`;
}
